src/Qir/Runtime/lib/QIR/QubitManagerRestrictedReuse.hpp

```diff
@@ -0,0 +1,210 @@
+
+        private: int* sharedQubitStatusArray; // Tracks allocation state of all qubits. Stores lists of free qubits.
+        private: int qubitCapacity = 0; // qubitCapacity is always equal to the array size.
+        private: CRestrictedReuseAreaStack freeQubitsInAreas; // Fresh Free Qubits are located in freeQubitsInAreas[0].FreeQubitsReuseAllowed
+        
+        public: CQubitManagerRestrictedReuse(
+            int initialQubitCapacity,
+            bool mayExtendCapacity = false,
+            bool encourageReuse = true);
+
+        // No complex scenarios for now. Don't need to support copying/moving.
+        public: CQubitManagerRestrictedReuse(const CQubitManagerRestrictedReuse&) = delete;
+        public: CQubitManagerRestrictedReuse& operator = (const CQubitManagerRestrictedReuse&) = delete;
+        public: ~CQubitManagerRestrictedReuse();
+        private: void EnsureCapacity(int requestedCapacity);
+
+        // Disable, disabled qubit count and checks
+
+        public: int DisabledQubitsCount = 0;
+        public: bool IsDisabled(int id);
+        public: bool IsDisabled(Qubit qubit);
+        // Disables a given qubit.
+        // Once a qubit is disabled it can never be "enabled" or reallocated.
+        public: void Disable(Qubit qubit);
+        // Disables a set of given qubits.
+        // Once a qubit is disabled it can never be "enabled" or reallocated.
+        public: void Disable(Qubit* qubitsToDisable, int qubitCount);
+
+        // Allocate, allocated qubit count and checks
+
+        public: int AllocatedQubitsCount = 0;
+        public: bool IsExplicitlyAllocated(long id);
+        public: void ChangeStatusToAllocated(long id);
+        // Allocates a qubit. Extend capacity if necessary and possible.
+        // Fails if the qubit cannot be allocated. 
+        public: Qubit Allocate();
+        // Allocates numToAllocate new qubits. Extend capacity if necessary and possible.
+        // Fails without allocating any qubits if the qubits cannot be allocated. 
+        // Returned array of qubits must be passed to Release to release qubits and free memory use for array.
+        public: Qubit* Allocate(long numToAllocate);
+
+        // Release, free qubit count and checks
+
+        public: long FreeQubitsCount = 0;
+        public: bool IsFree(long id);
+        public: bool IsFree(Qubit qubit);
+        public: void Release(long id);
+        // Releases a given qubit.
+        public: void Release(Qubit qubit);
+        // Releases an array of given qubits and deallocates array memory.
+        public: void Release(Qubit* qubitsToRelease, int qubitCount);
+
+        // Borrow (We treat borrowing as allocation currently)
+
+        public: Qubit Borrow();
+        public: Qubit* Borrow(long qubitCountToBorrow);
+
+        // Return (We treat returning as release currently)
+
+        public: void Return(Qubit qubit);
+        public: void Return(Qubit* qubitsToReturn, int qubitCountToReturn);
```

Comment:
Nit: If the _borrowing_ feature is optional then it might make sense to extract it into a separate interface (e.g. `IQubitLender`), either implemented or not implemented by the qubit manager classes.

Reply:
I don't think a separate interface is necessary. Borrowing functions are fully supported now, they just do not do borrowing. When user needs a qubit to borrow, we can't guarantee that such qubit exists; and if it does not, a new fresh qubit is allocated. So current implementation of borrowing is just takes this fallback approach. So it may be better to say that _true_ borrowing isn't supported. I removed remarks about borrowing in a few places.